[Bug]: Replace image flattening for RBD with creating pods for RBD clones

**Description ** 

In 4.22 RBD image flattening is not supported so changing test steps to create pods using RBD clones and start IO on them to fill cluster full ratio.

@skipif_external_mode
@magenta_squad
def test_clone_deletion_after_max_cluster_space_utilization(
    self, interface_type, pvc_clone_factory, threading_lock
):
    """
    Steps:
        1. Have a cluster with OCP + ODF internal mode
        2. Create cephfs + RBD pvcs with data
        3. Create clones of pvcs till cluster full ratio.
        4. Make the cluster out of full by increasing the full ratio.
        5. After the cluster is out of full state and IOs started , Try to delete clones.
        6. Clone deletion should be successful and should not give error messages.
    """

    # Creating the clones one by one and wait until they bound
    self.timeout = 1800
    logger.info(
        f"Start creating {self.num_of_clones} clones on {interface_type} PVC of size {self.pvc_size} GB."
    )

    # Create the initial set of clones
    self.create_clones(self.num_of_clones, pvc_clone_factory)

    # Maximum number of attempts to avoid indefinite looping
    max_attempts = 5
    attempt = 0

    # Verify if expected alerts are seen; if not, continue creating extra clones
    while attempt < max_attempts:
        logger.info(
            "Verify 'CephClusterCriticallyFull' ,CephOSDNearFull Alerts are seen "
        )
        expected_alerts = ["CephOSDNearFull", "CephOSDCriticallyFull"]
        prometheus = PrometheusAPI(threading_lock=threading_lock)
        sample = TimeoutSampler(
            timeout=180,
            sleep=10,
            func=prometheus.verify_alerts_via_prometheus,
            expected_alerts=expected_alerts,
            threading_lock=threading_lock,
        )

        if sample.wait_for_func_status(result=True):
            logger.info(
                "Expected alerts have been detected. Stopping clone creation."
            )
            break
        else:
            logger.info("Alerts not found yet. Creating extra clones...")
            # Continue creating more clones (e.g., in batches of 2)

          self.create_clones(2, start_num=len(self.clones_list))
E               TypeError: TestCloneDeletion.create_clones() missing 1 required positional argument: 'pvc_clone_factory'

--- a/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
+++ b/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
@@ -26,7 +26,7 @@ logger = logging.getLogger(__name__)
 @pytest.mark.parametrize(
     argnames=["interface_type"],
     argvalues=[
-        # pytest.param(constants.CEPHFILESYSTEM),
+        pytest.param(constants.CEPHFILESYSTEM),
         pytest.param(constants.CEPHBLOCKPOOL),
     ],
 )

--- a/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
+++ b/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
@@ -50,7 +50,29 @@ class TestCloneDeletion(E2ETest):
         """
 
         def teardown():
-            # change ceph full ratio to standard value
+            # Delete any clone pods that may have been created mid-test (RBD clones)
+            for clone_pod in self.clone_pods_list:
+                try:
+                    clone_pod.delete()
+                    clone_pod.ocp.wait_for_delete(
+                        resource_name=clone_pod.name, timeout=300
+                    )
+                except Exception:
+                    logger.warning(
+                        f"Failed to delete clone pod {clone_pod.name} during teardown"
+                    )
+
+            # Delete any clones (PVCs) that may have been created mid-test
+            for clone in self.clones_list:
+                try:
+                    clone.delete()
+                    clone.ocp.wait_for_delete(clone.name, timeout=300)
+                except Exception:
+                    logger.warning(
+                        f"Failed to delete clone {clone.name} during teardown"
+                    )
+
+            # Reset ceph full ratio to standard value
             change_ceph_full_ratio(85)
 
         request.addfinalizer(teardown)
@@ -159,7 +181,7 @@ class TestCloneDeletion(E2ETest):
         """
 
         logger.test_step(
-            f"Create {self.num_of_clones} clones of {self.pvc_size} GB "
+            f"Create {self.num_of_clones} clones of {self.pvc_size} GiB "
             f"on {interface_type} PVC to fill cluster"
         )
         self.timeout = 1800
@@ -219,8 +241,8 @@ class TestCloneDeletion(E2ETest):
                 clone_pod.ocp.wait_for_delete(resource_name=clone_pod.name, timeout=300)
             logger.info("All clone pods deleted successfully.")
 
-        logger.info(
-            f"Start deleting {self.num_of_clones} clones on {interface_type} PVC of size {self.pvc_size} Gi."
+        logger.test_step(
+            f"Delete {self.num_of_clones} clones on {interface_type} PVC of size {self.pvc_size} GiB"
         )
 
         for index, clone in enumerate(self.clones_list):

--- a/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
+++ b/tests/cross_functional/system_test/test_clone_deletion_after_max_cluster_space_utilization.py
@@ -5,7 +5,6 @@ from ocs_ci.ocs import constants
 from ocs_ci.framework.testlib import E2ETest
 from ocs_ci.utility.utils import TimeoutSampler
 from ocs_ci.helpers import helpers
-from ocs_ci.ocs.resources.pvc import flatten_image
 from ocs_ci.utility.prometheus import PrometheusAPI
 from ocs_ci.ocs.exceptions import TimeoutExpiredError
 from ocs_ci.framework.pytest_customization.marks import (
@@ -27,7 +26,7 @@ logger = logging.getLogger(__name__)
 @pytest.mark.parametrize(
     argnames=["interface_type"],
     argvalues=[
-        pytest.param(constants.CEPHFILESYSTEM),
+        # pytest.param(constants.CEPHFILESYSTEM),
         pytest.param(constants.CEPHBLOCKPOOL),
     ],
 )
@@ -60,6 +59,7 @@ class TestCloneDeletion(E2ETest):
 
         self.num_of_clones = 30
         self.clones_list = []
+        self.clone_pods_list = []
         self.interface_type = interface_type
 
         # Getting the total Storage capacity
@@ -111,21 +111,32 @@ class TestCloneDeletion(E2ETest):
         logger.info(f"IO finished on pod {self.pod_obj.name}")
 
     # Function to create clones
-    def create_clones(self, num_of_clones, pvc_clone_factory, start_num=0):
+    def create_clones(self, num_of_clones, pvc_clone_factory, pod_factory, start_num=0):
         for clone_num in range(start_num, start_num + num_of_clones):
             logger.info(f"Start creation of clone number {clone_num}.")
             cloned_pvc_obj = pvc_clone_factory(
                 self.pvc_obj, storageclass=self.pvc_obj.backed_sc, timeout=900
             )
+            cloned_pvc_obj.project = self.pvc_obj.project
             cloned_pvc_obj.reload()
 
             if self.interface_type == constants.CEPHBLOCKPOOL:
-                # flatten the image
-                flatten_image(cloned_pvc_obj)
+                # Write IO on the clone for RBD with Filesystem mode
                 logger.info(
-                    f"Clone with name {cloned_pvc_obj.name} of size {self.pvc_size}Gi was created."
+                    f"Clone with name {cloned_pvc_obj.name} of size {self.pvc_size}Gi was created. Writing IO to clone."
                 )
-                cloned_pvc_obj.reload()
+                clone_pod_obj = pod_factory(
+                    interface=constants.CEPHBLOCKPOOL,
+                    pvc=cloned_pvc_obj,
+                    raw_block_pv=False,
+                )
+                # Write IO using filesystem storage type
+                clone_pod_obj.run_io(
+                    size=self.filesize,
+                    io_direction="write",
+                    storage_type="fs",
+                )
+                self.clone_pods_list.append(clone_pod_obj)
             else:
                 logger.info(
                     f"Clone with name {cloned_pvc_obj.name} of size {self.pvc_size}Gi was created."
@@ -135,7 +146,7 @@ class TestCloneDeletion(E2ETest):
     @skipif_external_mode
     @magenta_squad
     def test_clone_deletion_after_max_cluster_space_utilization(
-        self, interface_type, pvc_clone_factory, threading_lock
+        self, interface_type, pvc_clone_factory, pod_factory, threading_lock
     ):
         """
         Steps:
@@ -154,7 +165,7 @@ class TestCloneDeletion(E2ETest):
         self.timeout = 1800
 
         # Create the initial set of clones
-        self.create_clones(self.num_of_clones, pvc_clone_factory)
+        self.create_clones(self.num_of_clones, pvc_clone_factory, pod_factory)
         logger.info(f"Successfully created {self.num_of_clones} initial clones")
 
         # Maximum number of attempts to avoid indefinite looping
@@ -182,7 +193,9 @@ class TestCloneDeletion(E2ETest):
             else:
                 logger.info("Alerts not found yet. Creating extra clones...")
                 # Continue creating more clones (e.g., in batches of 2)
-                self.create_clones(2, start_num=len(self.clones_list))
+                self.create_clones(
+                    2, pvc_clone_factory, pod_factory, start_num=len(self.clones_list)
+                )
                 attempt += 1
 
         if attempt == max_attempts:
@@ -195,9 +208,19 @@ class TestCloneDeletion(E2ETest):
         change_ceph_full_ratio(95)
         logger.info("Ceph full_ratio changed successfully")
 
-        logger.test_step(
-            f"Delete {len(self.clones_list)} clones on {interface_type} PVC "
-            f"of size {self.pvc_size} GB after cluster recovery"
+        # Delete clone pods first if they exist (for RBD clones)
+        if self.clone_pods_list:
+            logger.info(
+                f"Deleting {len(self.clone_pods_list)} clone pods for RBD clones."
+            )
+            for clone_pod in self.clone_pods_list:
+                logger.info(f"Deleting clone pod {clone_pod.name}")
+                clone_pod.delete()
+                clone_pod.ocp.wait_for_delete(resource_name=clone_pod.name, timeout=300)
+            logger.info("All clone pods deleted successfully.")
+
+        logger.info(
+            f"Start deleting {self.num_of_clones} clones on {interface_type} PVC of size {self.pvc_size} Gi."
         )
 
         for index, clone in enumerate(self.clones_list):

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -79,7 +79,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
     @pytest.fixture()
     def reduce_replication_delay_for_test(self, add_env_vars_to_noobaa_core):
         """
-        A function-scoped fixture to reduce the replication delay to one minute.
+        A function-scoped fixture to reduce the replication delay to two minutes.
         Changes will be automatically reverted after the test completes.
 
         Args:
@@ -87,7 +87,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         """
         from ocs_ci.ocs import constants as const
 
-        new_delay_in_milliseconds = 60 * 1000
+        new_delay_in_milliseconds = 2 * 60 * 1000
         new_env_var_tuples = [
             (const.BUCKET_REPLICATOR_DELAY_PARAM, new_delay_in_milliseconds),
         ]

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -79,7 +79,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
     @pytest.fixture()
     def reduce_replication_delay_for_test(self, add_env_vars_to_noobaa_core):
         """
-        A function-scoped fixture to reduce the replication delay to two minutes.
+        A function-scoped fixture to reduce the replication delay to one minute.
         Changes will be automatically reverted after the test completes.
 
         Args:
@@ -87,7 +87,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         """
         from ocs_ci.ocs import constants as const
 
-        new_delay_in_milliseconds = 2 * 60 * 1000
+        new_delay_in_milliseconds = 60 * 1000
         new_env_var_tuples = [
             (const.BUCKET_REPLICATOR_DELAY_PARAM, new_delay_in_milliseconds),
         ]
@@ -159,6 +159,10 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             prefix=prefix_site_1,
         )
         logger.info(f"Written objects: {written_random_objects}")
+
+        # Wait for replication cycle to complete
+        logger.info("Waiting 90 seconds for replication to complete...")
+        time.sleep(90)
 
         assert compare_bucket_object_list(
             mcg_obj_session, source_bucket_name, target_bucket_name

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -76,6 +76,23 @@ class TestMCGReplicationWithDisruptions(E2ETest):
     4) To verify that the Certain admin/disruptive operations do not impact the replication
     """
 
+    @pytest.fixture()
+    def reduce_replication_delay_for_test(self, add_env_vars_to_noobaa_core):
+        """
+        A function-scoped fixture to reduce the replication delay to one minute.
+        Changes will be automatically reverted after the test completes.
+
+        Args:
+            add_env_vars_to_noobaa_core (function): Function-scoped fixture to add env vars to noobaa-core pod
+        """
+        from ocs_ci.ocs import constants as const
+
+        new_delay_in_milliseconds = 60 * 1000
+        new_env_var_tuples = [
+            (const.BUCKET_REPLICATOR_DELAY_PARAM, new_delay_in_milliseconds),
+        ]
+        add_env_vars_to_noobaa_core(new_env_var_tuples)
+
     @pytest.mark.parametrize(
         argnames=["source_bucketclass", "target_bucketclass"],
         argvalues=[
@@ -114,6 +131,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         target_bucketclass,
         test_directory_setup,
         nodes,
+        reduce_replication_delay_for_test,
     ):
         logger.test_step("Setup uni-directional bucket replication")
         prefix_site_1 = "site1"

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -160,12 +160,8 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         )
         logger.info(f"Written objects: {written_random_objects}")
 
-        # Wait for replication cycle to complete
-        logger.info("Waiting 90 seconds for replication to complete...")
-        time.sleep(90)
-
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name
+            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
         )
         logger.info("Uni-directional bucket replication verified successfully")
 
@@ -192,9 +188,8 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             prefix=prefix_site_2,
         )
         logger.info(f"Written objects: {written_random_objects}")
-        time.sleep(90)
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name
+            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
         )
         logger.info("Bi-directional bucket replication verified successfully")
 
@@ -224,9 +219,8 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         )
         logger.info(f"Written objects: {written_random_objects}")
 
-        time.sleep(90)
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name
+            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
         )
         logger.info("All objects successfully recovered to target bucket on new write")
 
@@ -251,9 +245,8 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         wait_for_pods_to_be_running(
             pod_names=pod_names, namespace=config.ENV_DATA["cluster_namespace"]
         )
-        time.sleep(90)
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name
+            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
         )
         logger.info("Object replication verified successfully after RGW pod restart")
 
@@ -282,9 +275,8 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             namespace=config.ENV_DATA["cluster_namespace"], timeout=800
         )
         logger.info("Nodes rebooted successfully!!")
-        time.sleep(90)
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name
+            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
         )
         logger.info("Object replication verified successfully after cluster reboot")
 

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -76,23 +76,6 @@ class TestMCGReplicationWithDisruptions(E2ETest):
     4) To verify that the Certain admin/disruptive operations do not impact the replication
     """
 
-    @pytest.fixture()
-    def reduce_replication_delay_for_test(self, add_env_vars_to_noobaa_core):
-        """
-        A function-scoped fixture to reduce the replication delay to one minute.
-        Changes will be automatically reverted after the test completes.
-
-        Args:
-            add_env_vars_to_noobaa_core (function): Function-scoped fixture to add env vars to noobaa-core pod
-        """
-        from ocs_ci.ocs import constants as const
-
-        new_delay_in_milliseconds = 60 * 1000
-        new_env_var_tuples = [
-            (const.BUCKET_REPLICATOR_DELAY_PARAM, new_delay_in_milliseconds),
-        ]
-        add_env_vars_to_noobaa_core(new_env_var_tuples)
-
     @pytest.mark.parametrize(
         argnames=["source_bucketclass", "target_bucketclass"],
         argvalues=[
@@ -131,7 +114,6 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         target_bucketclass,
         test_directory_setup,
         nodes,
-        reduce_replication_delay_for_test,
     ):
         logger.test_step("Setup uni-directional bucket replication")
         prefix_site_1 = "site1"
@@ -161,7 +143,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         logger.info(f"Written objects: {written_random_objects}")
 
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
+            mcg_obj_session, source_bucket_name, target_bucket_name
         )
         logger.info("Uni-directional bucket replication verified successfully")
 
@@ -189,7 +171,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         )
         logger.info(f"Written objects: {written_random_objects}")
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
+            mcg_obj_session, source_bucket_name, target_bucket_name
         )
         logger.info("Bi-directional bucket replication verified successfully")
 
@@ -220,7 +202,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         logger.info(f"Written objects: {written_random_objects}")
 
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
+            mcg_obj_session, source_bucket_name, target_bucket_name
         )
         logger.info("All objects successfully recovered to target bucket on new write")
 
@@ -246,7 +228,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             pod_names=pod_names, namespace=config.ENV_DATA["cluster_namespace"]
         )
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
+            mcg_obj_session, source_bucket_name, target_bucket_name
         )
         logger.info("Object replication verified successfully after RGW pod restart")
 
@@ -276,7 +258,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         )
         logger.info("Nodes rebooted successfully!!")
         assert compare_bucket_object_list(
-            mcg_obj_session, source_bucket_name, target_bucket_name, timeout=1200
+            mcg_obj_session, source_bucket_name, target_bucket_name
         )
         logger.info("Object replication verified successfully after cluster reboot")
 

--- a/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
+++ b/tests/cross_functional/system_test/test_mcg_replication_with_disruptions.py
@@ -192,6 +192,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             prefix=prefix_site_2,
         )
         logger.info(f"Written objects: {written_random_objects}")
+        time.sleep(90)
         assert compare_bucket_object_list(
             mcg_obj_session, source_bucket_name, target_bucket_name
         )
@@ -223,6 +224,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         )
         logger.info(f"Written objects: {written_random_objects}")
 
+        time.sleep(90)
         assert compare_bucket_object_list(
             mcg_obj_session, source_bucket_name, target_bucket_name
         )
@@ -249,7 +251,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
         wait_for_pods_to_be_running(
             pod_names=pod_names, namespace=config.ENV_DATA["cluster_namespace"]
         )
-
+        time.sleep(90)
         assert compare_bucket_object_list(
             mcg_obj_session, source_bucket_name, target_bucket_name
         )
@@ -280,7 +282,7 @@ class TestMCGReplicationWithDisruptions(E2ETest):
             namespace=config.ENV_DATA["cluster_namespace"], timeout=800
         )
         logger.info("Nodes rebooted successfully!!")
-
+        time.sleep(90)
         assert compare_bucket_object_list(
             mcg_obj_session, source_bucket_name, target_bucket_name
         )

--- a/tests/cross_functional/system_test/test_object_expiration_system.py
+++ b/tests/cross_functional/system_test/test_object_expiration_system.py
@@ -415,7 +415,7 @@ class TestObjectExpirationSystemTest:
             )
             logger.info("Ceph health check passed after disruptions")
         except Exception as ex:
-            logger.warning(f"Ceph health check failed: {ex}")  
+            logger.warning(f"Ceph health check failed: {ex}")
 
         # check if the objects are expired
         sample_if_objects_expired()

--- a/tests/cross_functional/system_test/test_object_expiration_system.py
+++ b/tests/cross_functional/system_test/test_object_expiration_system.py
@@ -12,6 +12,8 @@ from ocs_ci.ocs.resources.mcg_lifecycle_policies import LifecyclePolicy, Expirat
 from ocs_ci.utility.retry import retry
 from ocs_ci.helpers.sanity_helpers import Sanity
 from ocs_ci.ocs import constants
+from ocs_ci.ocs.cluster import ceph_health_check, CephCluster
+from ocs_ci.framework import config
 from ocs_ci.framework.pytest_customization.marks import (
     system_test,
     magenta_squad,
@@ -401,7 +403,21 @@ class TestObjectExpirationSystemTest:
         wait_for_noobaa_pods_running(timeout=1200)
         logger.info("NooBaa DB pod node started, all NooBaa pods running")
 
-        logger.test_step("Verify object expiration after node disruptions")
+        logger.info("Waiting for NooBaa to stabilize after disruptions...")
+        ceph_cluster = CephCluster()
+        ceph_cluster.wait_for_noobaa_health_ok()
+        logger.info("NooBaa health check passed after disruptions")
+
+        # Perform ceph health check after disruptions
+        try:
+            ceph_health_check(
+                namespace=config.ENV_DATA["cluster_namespace"], tries=45, delay=60
+            )
+            logger.info("Ceph health check passed after disruptions")
+        except Exception as ex:
+            logger.warning(f"Ceph health check failed: {ex}")  
+
+        # check if the objects are expired
         sample_if_objects_expired()
         logger.info("Object expiration verified after node disruptions")
 

